rust/compaction/src/datafusion/functions/nonnull.rs

```diff
@@ -0,0 +1,1471 @@
+struct NonNullableAccumulator(Box<dyn Accumulator>);
+
+impl Accumulator for NonNullableAccumulator {
+    /// Performs a null check on the array.
+    ///
+    /// # Errors
+    /// Will produce an error if there are null values present.
+    fn update_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
+        self.0.update_batch(nullable_check(values)?)
+    }
+
+    fn evaluate(&mut self) -> Result<ScalarValue> {
+        self.0.evaluate()
+    }
+
+    fn size(&self) -> usize {
+        self.0.size() + std::mem::size_of::<Self>()
+    }
+
+    fn state(&mut self) -> Result<Vec<ScalarValue>> {
+        self.0.state()
+    }
+
+    fn merge_batch(&mut self, states: &[ArrayRef]) -> Result<()> {
+        self.0.merge_batch(states)
+    }
+
+    /// Performs a null check on the array.
+    ///
+    /// # Errors
+    /// Will produce an error if there are null values present.
+    fn retract_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
+        self.0.retract_batch(nullable_check(values)?)
+    }
+
+    fn supports_retract_batch(&self) -> bool {
+        self.0.supports_retract_batch()
+    }
+}
+
+/// Wraps the accumulation function with an [`Accumulator`] that checks the incoming
+/// arrays for nulls.
+struct NonNullableGroupsAccumulator(Box<dyn GroupsAccumulator>);
+
+impl GroupsAccumulator for NonNullableGroupsAccumulator {
+    /// Performs a null check on the array.
+    ///
+    /// # Errors
+    /// Will produce an error if there are null values present.
+    fn update_batch(
+        &mut self,
+        values: &[ArrayRef],
+        group_indices: &[usize],
+        opt_filter: Option<&BooleanArray>,
+        total_num_groups: usize,
+    ) -> Result<()> {
+        self.0.update_batch(
+            nullable_check(values)?,
+            group_indices,
+            opt_filter,
+            total_num_groups,
+        )
+    }
+
+    fn evaluate(&mut self, emit_to: EmitTo) -> Result<ArrayRef> {
+        self.0.evaluate(emit_to)
+    }
+
+    fn state(&mut self, emit_to: EmitTo) -> Result<Vec<ArrayRef>> {
+        self.0.state(emit_to)
+    }
+
+    fn merge_batch(
+        &mut self,
+        values: &[ArrayRef],
+        group_indices: &[usize],
+        opt_filter: Option<&BooleanArray>,
+        total_num_groups: usize,
+    ) -> Result<()> {
+        self.0
+            .merge_batch(values, group_indices, opt_filter, total_num_groups)
+    }
+
+    fn size(&self) -> usize {
+        self.0.size() + std::mem::size_of::<Self>()
+    }
+
+    /// Performs a null check on the array.
+    ///
+    /// # Errors
+    /// Will produce an error if there are null values present.
+    fn convert_to_state(
+        &self,
+        values: &[ArrayRef],
+        opt_filter: Option<&BooleanArray>,
+    ) -> Result<Vec<ArrayRef>> {
+        self.0.convert_to_state(nullable_check(values)?, opt_filter)
+    }
+
+    fn supports_convert_to_state(&self) -> bool {
+        self.0.supports_convert_to_state()
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I'm struggling to tell what these tests are asserting. What do they do?

Reply:
The tests with no explicit assertions are testing that the NonNullable wrapper structs are calling their underlying struct functions the correct number of times. The mocking framework will cause a test panic if the mock function is not called the correct number of times.